app/dashboard/static/js/app/view-tests-job-branch-kernel-plan.2020.1.js

```diff
@@ -0,0 +1,597 @@
+        }
+
+        batchOps = [];
+        results.forEach(createBatchOp)
+        deferred = request.post(
+            '/_ajax/batch', JSON.stringify({batch: batchOps}));
+
+        $.when(deferred)
+            .fail(error.error, getRegressionsFailed)
+            .done(getRegressionsDone);
+    }
+
+    function getLabResultsFailed() {
+        runsFailed();
+    }
+
+    function getLabResultsDone(response) {
+        updateLabs(response.result);
+    }
+
+    function getLabResults(results) {
+        var batchOps;
+        var deferred;
+        var params;
+
+        params = {
+            'job': gJob,
+            'kernel': gKernel,
+            'git_branch': gBranch,
+            'plan': gPlan,
+        };
+
+        function createBatchOp(lab) {
+            var qStr;
+
+            params['lab_name'] = lab;
+            qStr = URI.buildQuery(params);
+
+            batchOps.push({
+                method: 'GET',
+                operation_id: [lab, 'total'],
+                resource: 'count',
+                document: 'test_case',
+                query: qStr,
+            });
+
+            batchOps.push({
+                method: 'GET',
+                operation_id: [lab, 'success'],
+                resource: 'count',
+                document: 'test_case',
+                query: qStr + '&status=PASS',
+            });
+
+            batchOps.push({
+                method: 'GET',
+                operation_id: [lab, 'regressions'],
+                resource: 'count',
+                document: 'test_regression',
+                query: qStr,
+            });
+
+            batchOps.push({
+                method: 'GET',
+                operation_id: [lab, 'unknown'],
+                resource: 'count',
+                document: 'test_case',
+                query: qStr + '&status=FAIL&status=SKIP&regression_id=null',
+            });
+        }
+
+        batchOps = [];
+        Object.keys(gPanel.allLabs).forEach(createBatchOp);
+        deferred = request.post(
+            '/_ajax/batch', JSON.stringify({batch: batchOps}));
+
+        $.when(deferred)
+            .fail(error.error, getLabResultsFailed)
+            .done(getLabResultsDone);
+    }
+
+    function getRunsFailed() {
+        detailsFailed();
+        runsFailed();
+    }
+
+    function getRunsDone(response) {
+        updateDetails(response.result[0]);
+        updateRuns(response.result);
+        getRegressions(response.result);
+        getLabResults(response.result);
+    }
+
+    function getRuns() {
+        var data;
+        var deferred;
+
+        data = {
+            job: gJob,
+            git_branch: gBranch,
+            kernel: gKernel,
+            name: gPlan,
+            parent_id: 'null',
+            sort: 'device_type',
+            sort_order: 1,
+        };
+
+        deferred = request.get('/_ajax/test/group', data);
+        $.when(deferred)
+            .fail(error.error, getRunsFailed)
+            .done(getRunsDone);
```

Comment:
I see that you did a function called >getRunsDone, but it could be more simple, just calling the functions on done and passing as argument response and then into each function get what is necessary there. Is there any special reason to do it like you did? 

e.g

`.done(
                updateDetails,
                updateRuns,
                getRegressions,
                getLabResults
            );`

Reply:
I don't think it would actually be simpler to do that.  The principle is to separate the different levels in the code, first there is the request handling functions and then other functions that change the HTML content.  Functions like `getRunsDone` are clearly called when the request completes and they define what to do with the data by calling functions like `updateDetails`.  I think it makes it easier to maintain the code that way.

Reply:
sure